### ISSUE TITLE
Can't disable key-spacing.[before|after]Colon

### ISSUE THREAD

**Issue by zyklus** [eslint/eslint#3120]
The key-spacing rule seems to be overloaded so that I can't just disable the `beforeColon` rule without also disabling `align` and `afterColon`.  What I want:

```json
"key-spacing": [ 1, {
  "align"       : "colon",
  "beforeColon" : 0,
  "afterColon"  : true
} ]
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by zyklus** [comment]
eslint v0.24.1

Also note that this applies to _all_ compound rules.  I think you should be able to disable, for instance, `comma-spacing.before`, etc.

**Comment by IanVS** [comment]
Your configuration would only work if all property keys are exactly the same length.  That doesn't seem like a very common situation.  Or am I missing something?

**Comment by zyklus** [comment]
@IanVS - I'm saying I want to _disable_ the rule, as in I don't care if there is spacing before a colon or not.  e.g. I want these to be valid:

```json
{
  "foo"   : 42,
  "foobar": 43
}
```

But back to my main point, I don't think `align`, `beforeColon`, and `afterColon` should be "all or nothing".  The same goes for any other rule that has multiple sub-rules.  I think it should be possible to disable _any_ rule, at any level.

**Comment by nzakas** [comment]
@zyklus these are not subrules, they are related options. I think you just want `align: "colon" `, per the docs:

> The align option can also vertically align colons and values together. Whereas with "value" alignment, padding belongs right of the colon, with "colon" alignment, padding goes to the left of the colon. Except in the case of padding, it still respects beforeColon and afterColon. 

You can just leave off the options.

**Comment by ilyavolodin** [comment]
@zyklus If I understand what you are saying correctly, then what you want is:
```
"key-spacing": 0
```

That will disable the rule completely. Any of the rules in ESLint can be disabled by providing 0 as a first (or only) parameter.

**Comment by zyklus** [comment]
No, I want `align` to be enabled, but _not_ `beforeColon`

@nzakas -- sub-rules, related-options, it's just semantics :)  My point is that you can't enable `key-spacing.align` and disable `key-spacing.beforeColon`

Leaving out the `beforeColon` option, e.g. `"key-spacing": [1, {align: "colon"}]"` just sets it to its default value.

To clarify:

I _want_ to see `key-spacing.align` warnings.  I don't want to see `key-spacing.beforeColon` warnings, ever.

**Comment by nzakas** [comment]
I'm sorry, I'm having a really hard time understanding what you're describing. Can you please provide the exact ESLint output for your example code and point out what messages you don't want to see?

**Comment by zyklus** [comment]
I want these to pass:

```
foobar: 1
bar   : 2
```
```
foobar : 1
bar    : 2
```

but not these:

```
foobar: 1
bar: 2
```
```
foobar : 1
bar : 2
```

As I said, I just want to disable the `beforeColon` rule, that's all.  @ilyavolodin's suggestion to disable the `key-spacing` rule is _not_ what I want because it disables `align` as well.

And in a more general sense:

There are numerous "sub/related" rules like this.  I don't think there should _ever_ be rules tied together.

```js
key-spacing : // you can disable this
  align:        //
  beforeColon:  // You can't disable any of these individually
  afterColon:   //
```

**Comment by zyklus** [comment]
@nzakas -- I _think_ your confusion might stem from a mis-understanding of the way this rule works.  You said:

> You can just leave off the options.

If you leave off the options, they're still on.  They have default values that you can't disable.

**Comment by nzakas** [comment]
I'd like to ask again: can you please show your example code, plus your configuration, plus the output you're seeing? This info is really helpful for us to better understand what you're seeing.

@btmills can you chime in here?

**Comment by zyklus** [comment]
@nzakas -- I'm at a loss how to explain this more clearly.  I want to disable the `beforeColon` rule, that's it...

I didn't post the output because it's not useful since the rule produces the same warnings whether they're coming from having `align` or `beforeColon` set, but here:

source:
```js
var a = {
  a : 0,
  b : 0,

  c: 0,
  d: 0,

  e: 0,
  ff: 0,

  g : 0,
  hh : 0
};
```

using the following configs:

```json
{
  "rules": {
    "key-spacing" : [ 1, { "align": true } ]
  }
}
```
```
{
  "rules": {
    "key-spacing" : [ 1, { "align": true, "beforeColon": false } ]
  }
}
```

yields:

```bash
   2:2  warning  Extra space after key "a"    key-spacing # do not want -- from beforeColon: false
   3:2  warning  Extra space after key "b"    key-spacing # do not want -- from beforeColon: false
   8:2  warning  Missing space after key "e"  key-spacing
  12:2  warning  Extra space after key "hh"   key-spacing # do not want, want missing space on line 11, `g` instead
```

and this config:

```json
{
  "rules": {
    "key-spacing" : [ 1, { "align": true, "beforeColon": true } ]
  }
}
```

yields:

```bash
   5:2  warning  Missing space after key "c"   key-spacing # do not want -- from beforeColon: true
   6:2  warning  Missing space after key "d"   key-spacing # do not want -- from beforeColon: true
   8:2  warning  Missing space after key "e"   key-spacing
   9:2  warning  Missing space after key "ff"  key-spacing # do not want
  11:2  warning  Missing space after key "g"   key-spacing
```

**Comment by nzakas** [comment]
@zyklus I wouldn't ask for something that isn't helpful for my understanding. I'm fact, this cleared it up for me. What you want is not to turn beforeColon off, you want is to be "either". I'm assuming you want this to warn?

```js
var a = {
  g : 0,     // want to fail: mis-aligned
  hh: 0     // want to fail: mis-aligned
};
```

**Comment by btmills** [comment]
@nzakas I'll leave it to you to decide whether this case is sufficiently prevalent to justify its inclusion if somebody's willing to implement it. @zyklus do you have an idea for how this option would work without breaking backwards compatibility?

**Comment by zyklus** [comment]
@nzakas -- No, I really do want it to be off, as in the rule shouldn't ever trigger a warning.  I have no idea where our cross-talk is here?  The code you posted is fine and should not throw warnings

- I don't care if there is space before a colon or not
- I do care that colons aren't aligned

If this was 3 separate rules, instead of one "joined" rule, it wouldn't be an issue:

```json
"beforeColon" : 0,
"align"       : 1,
"afterColon"  : 1
```

**Comment by zyklus** [comment]
@btmills --

Yes, I propose being able to disable sub/related-rules.

```json
"key-align" : [ 1, { "align": true, "beforeColon": 0 } ]
```

I'd argue that this currently exists for "align" -- the default is, effectively:

```json
"key-align" : [ 1, { "align": 0, "beforeColon": false, "afterColon": true } ]
```

**Comment by btmills** [comment]
Instead of `0`, how about `"optional"` or `"either"`?

**Comment by zyklus** [comment]
My objection to "either" is that it seems @nzakas' interpretation of that is "true or false", i.e "enforce one or the other, as long as there is consistency".  I personally want to drop all enforcement.  I vote for `0` solely because it's the convention for "top-level" rules.  Having `0` some places and `"optional"` others would lead to confusion, no?

**Comment by nzakas** [comment]
I don't think we can go from true/false to adding a string value on top, nor do I think `0` would make any sense (again, these are options, not rules).

Unfortunately, I don't see a good way to support this change. We can't always accommodate every style everyone has, and so I think the best way to go would be for you to copy the existing rule into a custom rule and make the modifications you need.

**Comment by zyklus** [comment]
@nzakas -- what's wrong with `enum: [true, false, 'optional']`?

And I'm not suggesting that every _style_ be accommodated, I'm suggesting that every _option_ be disableable.

**Comment by nzakas** [comment]
@zyklus I'd appreciate it if you could refrain from lecturing me. We deal with hundreds of requests from people who want to tweak this or that rule, and unfortunately we can't always accommodate everyone's desires. That's exactly why we made ESLint pluggable, so you can create your own rules whenever we can't support what you want. That way, you're never blocked on getting what you want just because we disagree or are otherwise unable to implement your suggestion.

We have a responsibility to the larger ESLint community to make sure rules remain consistent, and that means we can't just start having boolean options become tri-state in one location. None of our decisions are arbitrary, they are always measured and with consideration for both our current user base and other related rules.

So yes, your user experience will be dictated by our design decisions, but not *solely* by our design decisions. In this case, the only way to support alignment properly was to have it be a part of this rule. We've looked at whether or not it would be possible to split that out from just spacing enforcement and were unable to figure out a good way to do it.

Maybe there's another option, and we're happy to keep discussing to see if we can come up with one, but only if you're willing to work with us and drop the accusatory tone whenever we disagree.

**Comment by zyklus** [comment]
@nzakas - My apologies, didn't mean to come off accusatory.  I was responding to what I perceived as a dismissal from your end and I've had more than a few conversations like this about flat out bugs in JSHint and was dismissively told "Don't care, change your code, won't fix, <closes issue>".

I definitely appreciate all the work you've put into eslint :)  It's a great tool and the philosophies behind it are definitely a huge step in the right direction and a boost to the community.  I'm not trying to alienate you :D  My concerns are that when you say it's pluggable, I'm sure you realize that the number of users willing to work with an AST is a fairly small subset (though perhaps there is more overlap in the users that can/are willing to and the users that _should_ be going against the grain and changing rules ;)).  It's therefore my opinion that ESLint's base rules should be as non-opinionated as possible.  That doesn't necessarily mean attempting to conform to everyone's style preferences, but I believe in cases like this decoupling rules is only a good thing.

Now it seems that your primary objection is that a) there is no good way to compatibly change the API here, b) a breaking API change affects too many users, and c) the number of users that care about this change is minimal -- well, I don't have a good argument to that ;)  But...

> that means we can't just start having boolean options become tri-state in one location

I'm not suggesting they become tri-state in one location.  I'm suggesting they become tri-state _everywhere_.  Or at least everywhere that there is an option that can't be disabled separately from it's rule.

The only other options I can even think of would be `enum: [ true, false, null/undefined (but set) ]`, but I'm guessing you object to that as well?  Or maybe something closer to: `[ 1, { afterColon: true, disable: [ 'beforeColon' ] } ] `

**Comment by nzakas** [comment]
We definitely don't want to dismiss concerns when they're brought up to us. The difference between JSHint and ESLint in this regard is that with ESLint, you actually have the option to create a custom rule to do exactly what you want. So when we can't or won't do what is asked, that doesn't mean *you* still can't do exactly what you want to do. When we suggest creating a custom rule, it's definitely not meant to be dismissive, but rather a way to say, "if this is really important to you, there's still a way to do it."

I'm afraid I don't see changing all options to be tri-state as viable. So far we only have one use case where this is problematic (this one), and I don't see that as firm grounds to create such a disruptive change.

Looking at this again, I wonder if what we're really after is a new option called `alignSpacing` that is `true` by default and determines whether `beforeColon` and `afterColon` are used. @btmills would that make sense?

**Comment by ilyavolodin** [comment]
Closing this issue since there hasn't been any movement on this issue in a while, and decision hasn't been reached. Feel free to reopen it if you have any new proposal on how to achieve this.
